pkg/container/docker_runner.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2022 Chainguard, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package container
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+)
+
+type DKRunner struct {
+	Runner
+}
+
+// DockerRunner returns a Docker Runner implementation.
+func DockerRunner() Runner {
+	return &DKRunner{}
+}
+
+// StartPod starts a pod for supporting a Docker task, if
+// necessary.
+func (dk *DKRunner) StartPod(cfg *Config) error {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	mounts := []mount.Mount{}
+	for _, bind := range cfg.Mounts {
+		mounts = append(mounts, mount.Mount{
+			Type:   mount.TypeBind,
+			Source: bind.Source,
+			Target: bind.Destination,
+		})
+	}
+
+	hostConfig := &container.HostConfig{
+		Mounts: mounts,
+	}
+
+	ctx := context.Background()
+	resp, err := cli.ContainerCreate(ctx, &container.Config{
+		Image: cfg.ImgDigest,
+		Cmd:   []string{"/bin/sh", "-c", "while true; do sleep 5; done"},
+		Tty:   false,
+	}, hostConfig, nil, nil, "")
+	if err != nil {
+		return err
+	}
+
+	if err := cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
+		return err
+	}
+
+	cfg.PodID = resp.ID
+	cfg.Logger.Printf("pod %s started.", cfg.PodID)
+
+	return nil
+}
+
+// TerminatePod terminates a pod for supporting a Docker task,
+// if necessary.
+func (dk *DKRunner) TerminatePod(cfg *Config) error {
+	if cfg.PodID == "" {
+		return fmt.Errorf("pod not running")
+	}
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	ctx := context.Background()
+	if err := cli.ContainerRemove(ctx, cfg.PodID, types.ContainerRemoveOptions{
+		Force: true,
+	}); err != nil {
+		return err
+	}
+
+	cfg.Logger.Printf("pod %s terminated.", cfg.PodID)
+
+	return nil
+}
+
+// Run runs a Docker task given a Config and command string.
+func (dk *DKRunner) Run(cfg *Config, args ...string) error {
```

Comment:
In the case of doing it this way with Docker, it is so that we can encapsulate state.  In Kubernetes, we would just spawn new containers for each build task, most likely.

Reply:
However we want to do it.  What Melange calls a "pod" is just the state encapsulation.  We can also rename that terminology if it is too confusing :)